Comments feature

SofiaVega opened the pull request:
Add fetch function to request a message from the servlet.

SofiaVega committed:
Changed message displayed at /data, Step 1 for walkthrough 3

SofiaVega committed:
Add greeting button, modify script
Script now fetches a message from DataServlet.java

SofiaVega committed:
Correct styling.

SofiaVega committed:
Step 3: Add comments to page from JSON.

SofiaVega committed:
Move comments to bottom of page, 'getJson()' now runs on load, remove 'Start' button.

SofiaVega committed:
Add 'new-comment' form
Update pom.xml file with project ID, add doPost method in servlet, redirect to index.html after comment has been posted.

SofiaVega committed:
Step 5: Use datastore to store and load comments
Add AppEngine dependency to pom.xml, modify DataServlet to store and load comments from Datastore

SofiaVega committed:
Add 'Delete comments' button
Add input to request the number of comments shown, default is three. Add new servlet to delete comments.

SofiaVega committed:
Add 'username' element to the comments form
Comments now show username as well as the comment (default value is Guest)
Change styling

livvielin commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,81 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-
+  int maxComments=3;
```
nit: I'd put whitespace so that it looks like `int maxComments = 3` for things like this and other places below. See https://google.github.io/styleguide/javaguide.html#s4.6-whitespace for additional details

Since this appears to be a constant, it should probably look like `static final int MAX_COMMENTS = 3` according to https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

Also, should this be a private variable?

SofiaVega replied:
Sorry, I'll the name because this is not a constant (not a maximum number). I did make it a private variable.

SofiaVega replied:
*I'll change the name

livvielin commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,81 @@
-
+  int maxComments=3;
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  /* Returns an array of comments in JSON 
```
nit: Javadoc should look a little more like https://google.github.io/styleguide/javaguide.html#s7.1.1-javadoc-multi-line. For this, it can probably just be a single line `/** Returns an array of comments in JSON with size maxComments. */`

livvielin commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,81 @@
+    ArrayList<String[]> comments = new ArrayList<String[]>();
+    int index=0;
+    for (Entity entity : results.asIterable()) {
+      if(index<maxComments){
```
I think this code is a bit hard to parse at first since it's unclear what's a "Comment" versus an "Entity". It seems like what you're doing is essentially getting the first three entities and transforming them into comments? (But let me know if I'm misinterpreting.) So maybe instead of this, you could just slice the results from 0 to MAX_COMMENTS, then iterate through the sliced version to transform them into comments?

As an additional note, it might help to comment this bit of code.

Feel free to push back on this if it doesn't make much sense.

SofiaVega replied:
You're right, I moved this inside a new function and it now iterates only over a sliced version of the results. I think it makes more sense now.

livvielin commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,81 @@
+    
+  }
+
+  /* Adds the comment and username from the 'new-comment' and 'username' 
```
nit: see the Javadoc comment above, specifically for multi-line formatting

livvielin commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,81 @@
+      }
+      catch (NumberFormatException e)
+      {
+        //maxComments = 3;
```
note: reminder to update this / put something here

livvielin commented on portfolio/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -0,0 +1,11 @@
+<web-app xmlns="http://xmlns.jcp.org/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd" version="3.1">
```
Should probably use https if possible

livvielin commented on portfolio/src/main/webapp/style.css:
```diff
@@ -61,3 +61,70 @@ h1 {
+  padding: 15px;
+  margin-bottom: 15px;
+  height: 50px;
+  border: solid lightgray
```
nit: alphabetize declarations in this file

livvielin commented on portfolio/src/main/webapp/script.js:
```diff
@@ -27,3 +27,50 @@ function addRandomFact() {
+}
+
+/** Adds a random quote to the DOM. */
+function addGreetingToDom(greeting) {
```
Note that JSDoc includes params: https://google.github.io/styleguide/jsguide.html#jsdoc

livvielin reviewed (commented):
Added some initial comments to get you started. I think this is probably a good size for a PR, perhaps even on the long end, so I'd recommend a separate PR for follow-ups.

SofiaVega committed:
Fix style
Delete functions in script.js that are not used anymore, split functions in script.js
and add a new method in DataServlet.java. Also added commens in all files and completed JSdoc
Add LoginServlet.java to implement login feature (for next pull request, it's not complete yet)

livvielin commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -21,8 +21,11 @@
+import com.google.appengine.api.datastore.FetchOptions;
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.List;
```
I think the order of this and the one below should be swapped? Do you have a linter that will format for you?

livvielin commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -31,64 +34,62 @@
-      }
-      index++;
-    }
+    ArrayList<String[]> comments = getComments();
```
Not necessarily for this PR, but it could be helpful to create a Comment class instead of just turning it into a String

livvielin reviewed (approved):
changes lgtm, left some minor comments
